sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/iceberg/IcebergMetastore.java

```diff
@@ -0,0 +1,154 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.sql.meta.provider.iceberg;
+
+import static org.apache.beam.sdk.util.Preconditions.checkStateNotNull;
+import static org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.base.Preconditions.checkArgument;
+
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.beam.sdk.extensions.sql.TableUtils;
+import org.apache.beam.sdk.extensions.sql.impl.TableName;
+import org.apache.beam.sdk.extensions.sql.meta.BeamSqlTable;
+import org.apache.beam.sdk.extensions.sql.meta.Table;
+import org.apache.beam.sdk.extensions.sql.meta.provider.TableProvider;
+import org.apache.beam.sdk.extensions.sql.meta.store.InMemoryMetaStore;
+import org.apache.beam.sdk.io.iceberg.IcebergCatalogConfig;
+import org.apache.beam.sdk.io.iceberg.IcebergCatalogConfig.IcebergTableInfo;
+import org.apache.beam.sdk.io.iceberg.TableAlreadyExistsException;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.annotations.VisibleForTesting;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.ImmutableMap;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class IcebergMetastore extends InMemoryMetaStore {
+  private static final Logger LOG = LoggerFactory.getLogger(IcebergMetastore.class);
+  @VisibleForTesting final IcebergCatalogConfig catalogConfig;
+  private final Map<String, Table> cachedTables = new HashMap<>();
+  private final String database;
+
+  public IcebergMetastore(String db, IcebergCatalogConfig catalogConfig) {
+    this.database = db;
+    this.catalogConfig = catalogConfig;
+  }
+
+  @Override
+  public String getTableType() {
+    return "iceberg";
+  }
+
+  @Override
+  public void createTable(Table table) {
+    if (!table.getType().equals("iceberg")) {
+      getProvider(table.getType()).createTable(table);
+    } else {
+      String identifier = getIdentifier(table);
+      try {
+        catalogConfig.createTable(identifier, table.getSchema(), table.getPartitionFields());
+      } catch (TableAlreadyExistsException e) {
+        LOG.info(
+            "Iceberg table '{}' already exists at location '{}'.", table.getName(), identifier);
+      }
+    }
+    cachedTables.put(table.getName(), table);
+  }
+
+  @Override
+  public void dropTable(String tableName) {
+    String identifier = getIdentifier(tableName);
+    if (catalogConfig.dropTable(identifier)) {
+      LOG.info("Dropped table '{}' (path: '{}').", tableName, identifier);
+    } else {
+      LOG.info(
+          "Ignoring DROP TABLE call for '{}' (path: '{}') because it does not exist.",
+          tableName,
+          identifier);
+    }
+    cachedTables.remove(tableName);
+  }
+
+  @Override
+  public Map<String, Table> getTables() {
+    for (String id : catalogConfig.listTables(database)) {
```

Comment:
@ahmedabu98 How frequently we call this method ? I feel this is very inefficient.